be/src/olap/base_compaction.cpp

```diff
@@ -16,511 +16,151 @@
-            << ", cumulative_base_ratio=" << cumulative_base_ratio
-            << ", policy_ratio=" << base_cumulative_delta_ratio;
-        return true;
-    }
-
-    // 触发条件3：距离上一次进行base compaction已经超过设定的间隔时间
-    const uint32_t interval_since_last_operation = config::base_compaction_interval_seconds_since_last_operation;
-    int64_t interval_since_last_be = time(NULL) - base_creation_time;
-    if (interval_since_last_be > interval_since_last_operation) {
+                  << ", cumualtive_total_size=" << cumulative_total_size
+                  << ", base_size=" << base_size
+                  << ", cumulative_base_ratio=" << cumulative_base_ratio
+                  << ", policy_ratio=" << base_cumulative_delta_ratio;
+        return OLAP_SUCCESS;
+    }
+
+    // 3. the interval since last base compaction reachs the threshold
+    int64_t base_creation_time = _input_rowsets[0]->creation_time();
+    int64_t interval_threshold = config::base_compaction_interval_seconds_since_last_operation;
+    int64_t interval_since_last_base_compaction = time(NULL) - base_creation_time;
+    if (interval_since_last_base_compaction > interval_threshold) {
         LOG(INFO) << "satisfy the base compaction policy. tablet=" << _tablet->full_name()
-            << ", interval_since_last_be=" << interval_since_last_be
-            << ", policy_interval=" << interval_since_last_operation;
-        return true;
+                  << ", interval_since_last_base_compaction=" << interval_since_last_base_compaction 
+                   << ", interval_threshold=" << interval_threshold;
+        return OLAP_SUCCESS;
     }
 
-    VLOG(3) << "don't satisfy the base compaction policy. tablet=" << _tablet->full_name()
-        << ", cumulative_files_number=" << candidate_versions->size() - 1
-        << ", cumulative_base_ratio=" << cumulative_base_ratio
-        << ", interval_since_last_be=" << interval_since_last_be;
-
-    return false;
+    LOG(INFO) << "don't satisfy the base compaction policy. tablet=" << _tablet->full_name()
+              << ", num_cumulative_rowsets=" << _input_rowsets.size() - 1
+              << ", cumulative_base_ratio=" << cumulative_base_ratio
+              << ", interval_since_last_base_compaction=" << interval_since_last_base_compaction;
+    return OLAP_ERR_BE_NO_SUITABLE_VERSION;
 }
 
-OLAPStatus BaseCompaction::_do_base_compaction(VersionHash new_base_version_hash,
-                                               const vector<RowsetSharedPtr>& rowsets) {
-    OlapStopWatch watch;
-    vector<RowsetReaderSharedPtr> rs_readers;
-    for (auto& rowset : rowsets) {
-        RowsetReaderSharedPtr rs_reader(rowset->create_reader());
-        if (rs_reader == nullptr) {
-            LOG(WARNING) << "rowset create reader failed. rowset:" <<  rowset->rowset_id();
-            return OLAP_ERR_ROWSET_CREATE_READER;
-        }
-        rs_readers.push_back(rs_reader);
+OLAPStatus BaseCompaction::do_compaction() {
```

Comment:
I think it's better to encapsulate all the following logic into base class.

Reply:
I do not think so.
Different Compaction can has different logic. 
If I want to change base compaction to capture partial rowsets, I can do it in this function.

Reply:
most of this function's logic is the same with cumulative compaction. 
1. pick rowset
2. check rowset if continue
3. create readers and writer
4. check rows if match
5. write metadata

Now, there are two same logic in two files, I think we can make it better.